File: sdk/python/atb/poa.py
"""Proof-of-Authorization (PoA) data structures and builder."""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Literal

import jwt

from atb.exceptions import ValidationError


@dataclass
class AccountableParty:
    """The human or organizational entity legally accountable for the action."""

    type: Literal["user", "service_account", "org_unit", "role"]
    id: str
    display_name: str | None = None

    def to_dict(self) -> dict:
        """Convert to dictionary, excluding None values."""
        result = {"type": self.type, "id": self.id}
        if self.display_name:
            result["display_name"] = self.display_name
        return result


@dataclass
class DualControl:
    """Dual-control / four-eyes approval metadata."""

    required: bool
    approver: AccountableParty | None = None
    approved_at: str | None = None

    def to_dict(self) -> dict:
        """Convert to dictionary."""
        result = {"required": self.required}
        if self.approver:
            result["approver"] = self.approver.to_dict()
        if self.approved_at:
            result["approved_at"] = self.approved_at
        return result


@dataclass
class LegalGrounding:
    """Legal grounding for a PoA mandate."""

    jurisdiction: str
    accountable_party: AccountableParty
    approval_ref: str | None = None
    dual_control: DualControl | None = None
    regulation_refs: list[str] = field(default_factory=list)
    retention_days: int | None = None

    def to_dict(self) -> dict:
        """Convert to dictionary."""
        result = {
            "jurisdiction": self.jurisdiction,
            "accountable_party": self.accountable_party.to_dict(),
        }
        if self.approval_ref:
            result["approval_ref"] = self.approval_ref
        if self.dual_control:
            result["dual_control"] = self.dual_control.to_dict()
        if self.regulation_refs:
            result["regulation_refs"] = self.regulation_refs
        if self.retention_days:
            result["retention_days"] = self.retention_days
        return result


@dataclass
class PoA:
    """Proof-of-Authorization mandate token."""

    # Required fields
    sub: str  # SPIFFE ID of the agent
    act: str  # Action identifier (e.g., "sap.vendor.change")
    con: dict[str, Any]  # Constraints - params and constraints
    leg: LegalGrounding  # Legal grounding
    iat: int  # Issued at (Unix timestamp)
    exp: int  # Expiration (Unix timestamp)
    jti: str  # JWT ID for replay protection

    # Optional fields
    iss: str | None = None  # Issuer SPIFFE ID
    aud: str | list[str] | None = None  # Audience SPIFFE ID(s)

    def to_dict(self) -> dict:
        """Convert to dictionary for JWT encoding."""
        result = {
            "sub": self.sub,
            "act": self.act,
            "con": self.con,
            "leg": self.leg.to_dict(),
            "iat": self.iat,
            "exp": self.exp,
            "jti": self.jti,
        }
        if self.iss:
            result["iss"] = self.iss
        if self.aud:
            result["aud"] = self.aud
        return result

    def to_jwt(self, private_key: str, algorithm: str = "ES256") -> str:
        """Sign and encode the PoA as a JWT.

        Args:
            private_key: PEM-encoded private key for signing.
            algorithm: JWT signing algorithm (default: ES256).

        Returns:
            Signed JWT string.
        """
        return jwt.encode(self.to_dict(), private_key, algorithm=algorithm)

    def is_expired(self) -> bool:
        """Check if the PoA has expired."""
        return time.time() > self.exp

    def validate(self) -> None:
        """Validate the PoA structure.

        Raises:
            ValidationError: If validation fails.
        """
        # Check SPIFFE ID format
        if not self.sub.startswith("spiffe://"):
            raise ValidationError("Subject must be a valid SPIFFE ID", field="sub")

        # Check action format (dot-separated namespace)
        if "." not in self.act:
            raise ValidationError(
                "Action must be dot-separated (e.g., 'sap.vendor.change')",
                field="act",
            )

        # Check timestamps
        if self.exp <= self.iat:
            raise ValidationError("Expiration must be after issued time", field="exp")

        # Check JTI
        if len(self.jti) < 8:
            raise ValidationError("JTI must be at least 8 characters", field="jti")

        # Check legal grounding
        if not self.leg.jurisdiction:
            raise ValidationError("Jurisdiction is required", field="leg.jurisdiction")

    @classmethod
    def from_jwt(cls, token: str, public_key: str, algorithms: list[str] | None = None) -> PoA:
        """Decode and verify a JWT to create a PoA.

        Args:
            token: JWT string to decode.
            public_key: PEM-encoded public key for verification.
            algorithms: Allowed algorithms (default: ["ES256"]).

        Returns:
            PoA instance.

        Raises:
            ValidationError: If token is invalid or expired.
        """
        algorithms = algorithms or ["ES256"]
        try:
            payload = jwt.decode(token, public_key, algorithms=algorithms)
        except jwt.ExpiredSignatureError:
            raise ValidationError("Token has expired") from None
        except jwt.InvalidTokenError as e:
            raise ValidationError(f"Invalid token: {e}") from e

        leg_data = payload["leg"]
        accountable_party = AccountableParty(
            type=leg_data["accountable_party"]["type"],
            id=leg_data["accountable_party"]["id"],
            display_name=leg_data["accountable_party"].get("display_name"),
        )

        dual_control = None
        if "dual_control" in leg_data:
            dc = leg_data["dual_control"]
            approver = None
            if "approver" in dc:
                approver = AccountableParty(
                    type=dc["approver"]["type"],
                    id=dc["approver"]["id"],
                    display_name=dc["approver"].get("display_name"),
                )
            dual_control = DualControl(
                required=dc["required"],
                approver=approver,
                approved_at=dc.get("approved_at"),
            )

        leg = LegalGrounding(
            jurisdiction=leg_data["jurisdiction"],
            accountable_party=accountable_party,
            approval_ref=leg_data.get("approval_ref"),
            dual_control=dual_control,
            regulation_refs=leg_data.get("regulation_refs", []),
            retention_days=leg_data.get("retention_days"),
        )

        return cls(
            sub=payload["sub"],
            act=payload["act"],
            con=payload["con"],
            leg=leg,
            iat=payload["iat"],
            exp=payload["exp"],
            jti=payload["jti"],
            iss=payload.get("iss"),
            aud=payload.get("aud"),
        )


class PoABuilder:
    """Builder for creating PoA mandates with a fluent interface."""

    def __init__(self) -> None:
        self._sub: str | None = None
        self._act: str | None = None
        self._params: dict[str, Any] = {}
        self._constraints: dict[str, Any] = {}
        self._leg: LegalGrounding | None = None
        self._ttl: int = 300  # Default 5 minutes
        self._iss: str | None = None
        self._aud: str | list[str] | None = None

    def for_agent(self, spiffe_id: str) -> PoABuilder:
        """Set the subject (agent SPIFFE ID).

        Args:
            spiffe_id: SPIFFE ID of the agent (e.g., "spiffe://atb.example/agent/copilot").
        """
        self._sub = spiffe_id
        return self

    def action(self, action: str) -> PoABuilder:
        """Set the action to authorize.

        Args:
            action: Action identifier (e.g., "sap.vendor.change").
        """
        self._act = action
        return self

    def with_params(self, **params: Any) -> PoABuilder:
        """Add action parameters.

        Args:
            **params: Key-value pairs for action parameters.
        """
        self._params.update(params)
        return self

    def with_constraint(self, name: str, value: Any) -> PoABuilder:
        """Add a constraint.

        Args:
            name: Constraint name (e.g., "liability_cap").
            value: Constraint value.
        """
        self._constraints[name] = value
        return self

    def with_constraints(self, **constraints: Any) -> PoABuilder:
        """Add multiple constraints.

        Args:
            **constraints: Key-value pairs for constraints.
        """
        self._constraints.update(constraints)
        return self

    def legal(
        self,
        jurisdiction: str,
        accountable_party: AccountableParty,
        approval_ref: str | None = None,
        dual_control: DualControl | None = None,
        regulation_refs: list[str] | None = None,
        retention_days: int | None = None,
    ) -> PoABuilder:
        """Set legal grounding.

        Args:
            jurisdiction: ISO 3166-1 alpha-2 country code or "GLOBAL".
            accountable_party: The human/org legally accountable.
            approval_ref: External approval reference (e.g., "SNOW-CHG0012345").
            dual_control: Dual-control approval metadata.
            regulation_refs: Regulatory frameworks (e.g., ["NIS2", "SOX"]).
            retention_days: Audit retention period in days.
        """
        self._leg = LegalGrounding(
            jurisdiction=jurisdiction,
            accountable_party=accountable_party,
            approval_ref=approval_ref,
            dual_control=dual_control,
            regulation_refs=regulation_refs or [],
            retention_days=retention_days,
        )
        return self

    def ttl(self, seconds: int) -> PoABuilder:
        """Set time-to-live (default: 300 seconds / 5 minutes).

        Args:
            seconds: TTL in seconds (recommended max: 300).
        """
        self._ttl = seconds
        return self

    def issuer(self, spiffe_id: str) -> PoABuilder:
        """Set the issuer SPIFFE ID.

        Args:
            spiffe_id: SPIFFE ID of the issuing service.
        """
        self._iss = spiffe_id
        return self

    def audience(self, spiffe_id: str | list[str]) -> PoABuilder:
        """Set the audience SPIFFE ID(s).

        Args:
            spiffe_id: SPIFFE ID or list of IDs for the broker(s).
        """
        self._aud = spiffe_id
        return self

    def build(self) -> PoA:
        """Build the PoA mandate.

        Returns:
            PoA instance.

        Raises:
            ValidationError: If required fields are missing.
        """
        if not self._sub:
            raise ValidationError("Subject (agent SPIFFE ID) is required")
        if not self._act:
            raise ValidationError("Action is required")
        if not self._leg:
            raise ValidationError("Legal grounding is required")

        now = int(time.time())
        poa = PoA(
            sub=self._sub,
            act=self._act,
            con={"params": self._params, "constraints": self._constraints},
            leg=self._leg,
            iat=now,
            exp=now + self._ttl,
            jti=f"poa_{uuid.uuid4().hex}",
            iss=self._iss,
            aud=self._aud,
        )
        poa.validate()
        return poa

File: sdk/go/poa.go
package atb

import (
	"crypto/ecdsa"
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
)

// AccountableParty represents the human or organization legally accountable.
type AccountableParty struct {
	Type        string `json:"type"`         // "user", "service_account", "org_unit", "role"
	ID          string `json:"id"`           // Unique identifier
	DisplayName string `json:"display_name,omitempty"` // Human-readable name
}

// DualControl represents dual-control / four-eyes approval metadata.
type DualControl struct {
	Required   bool              `json:"required"`
	Approver   *AccountableParty `json:"approver,omitempty"`
	ApprovedAt string            `json:"approved_at,omitempty"` // ISO 8601 timestamp
}

// LegalGrounding represents the legal basis for a PoA mandate.
type LegalGrounding struct {
	Jurisdiction     string            `json:"jurisdiction"`      // ISO 3166-1 alpha-2 or "GLOBAL"
	AccountableParty AccountableParty  `json:"accountable_party"` // Who is legally responsible
	ApprovalRef      string            `json:"approval_ref,omitempty"` // External approval reference
	DualControl      *DualControl      `json:"dual_control,omitempty"` // Dual-control metadata
	RegulationRefs   []string          `json:"regulation_refs,omitempty"` // e.g., ["NIS2", "SOX"]
	RetentionDays    int               `json:"retention_days,omitempty"` // Audit retention period
}

// Constraints represents action parameters and policy constraints.
type Constraints struct {
	Params      map[string]any `json:"params,omitempty"`      // Action parameters
	Constraints map[string]any `json:"constraints,omitempty"` // Policy constraints
}

// PoA represents a Proof-of-Authorization mandate.
type PoA struct {
	Sub string         `json:"sub"` // Agent SPIFFE ID
	Act string         `json:"act"` // Action identifier
	Con Constraints    `json:"con"` // Constraints
	Leg LegalGrounding `json:"leg"` // Legal grounding
	Iat int64          `json:"iat"` // Issued at (Unix timestamp)
	Exp int64          `json:"exp"` // Expiration (Unix timestamp)
	Jti string         `json:"jti"` // JWT ID for replay protection
	Iss string         `json:"iss,omitempty"` // Issuer SPIFFE ID
	Aud []string       `json:"aud,omitempty"` // Audience SPIFFE ID(s)
}

// IsExpired checks if the PoA has expired.
func (p *PoA) IsExpired() bool {
	return time.Now().Unix() > p.Exp
}

// Validate checks the PoA structure for validity.
func (p *PoA) Validate() error {
	// Check SPIFFE ID format
	if len(p.Sub) < 10 || p.Sub[:9] != "spiffe://" {
		return fmt.Errorf("%w: subject must be a valid SPIFFE ID", ErrValidation)
	}

	// Check action format
	hasNamespace := false
	for _, c := range p.Act {
		if c == '.' {
			hasNamespace = true
			break
		}
	}
	if !hasNamespace {
		return fmt.Errorf("%w: action must be dot-separated (e.g., 'sap.vendor.change')", ErrValidation)
	}

	// Check timestamps
	if p.Exp <= p.Iat {
		return fmt.Errorf("%w: expiration must be after issued time", ErrValidation)
	}

	// Check JTI
	if len(p.Jti) < 8 {
		return fmt.Errorf("%w: jti must be at least 8 characters", ErrValidation)
	}

	// Check legal grounding
	if p.Leg.Jurisdiction == "" {
		return fmt.Errorf("%w: jurisdiction is required", ErrValidation)
	}
	if p.Leg.AccountableParty.ID == "" {
		return fmt.Errorf("%w: accountable party ID is required", ErrValidation)
	}

	return nil
}

// ToJWT signs and encodes the PoA as a JWT.
func (p *PoA) ToJWT(privateKey *ecdsa.PrivateKey) (string, error) {
	claims := jwt.MapClaims{
		"sub": p.Sub,
		"act": p.Act,
		"con": map[string]any{
			"params":      p.Con.Params,
			"constraints": p.Con.Constraints,
		},
		"leg": map[string]any{
			"jurisdiction": p.Leg.Jurisdiction,
			"accountable_party": map[string]any{
				"type":         p.Leg.AccountableParty.Type,
				"id":           p.Leg.AccountableParty.ID,
				"display_name": p.Leg.AccountableParty.DisplayName,
			},
		},
		"iat": p.Iat,
		"exp": p.Exp,
		"jti": p.Jti,
	}

	if p.Iss != "" {
		claims["iss"] = p.Iss
	}
	if len(p.Aud) > 0 {
		claims["aud"] = p.Aud
	}
	if p.Leg.ApprovalRef != "" {
		claims["leg"].(map[string]any)["approval_ref"] = p.Leg.ApprovalRef
	}
	if p.Leg.DualControl != nil {
		claims["leg"].(map[string]any)["dual_control"] = p.Leg.DualControl
	}
	if len(p.Leg.RegulationRefs) > 0 {
		claims["leg"].(map[string]any)["regulation_refs"] = p.Leg.RegulationRefs
	}
	if p.Leg.RetentionDays > 0 {
		claims["leg"].(map[string]any)["retention_days"] = p.Leg.RetentionDays
	}

	token := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
	return token.SignedString(privateKey)
}

// PoABuilder provides a fluent interface for building PoA mandates.
type PoABuilder struct {
	sub         string
	act         string
	params      map[string]any
	constraints map[string]any
	leg         *LegalGrounding
	ttl         time.Duration
	iss         string
	aud         []string
}

// NewPoABuilder creates a new PoA builder with default values.
func NewPoABuilder() *PoABuilder {
	return &PoABuilder{
		params:      make(map[string]any),
		constraints: make(map[string]any),
		ttl:         5 * time.Minute,
	}
}

// ForAgent sets the subject (agent SPIFFE ID).
func (b *PoABuilder) ForAgent(spiffeID string) *PoABuilder {
	b.sub = spiffeID
	return b
}

// Action sets the action to authorize.
func (b *PoABuilder) Action(action string) *PoABuilder {
	b.act = action
	return b
}

// WithParams adds action parameters.
func (b *PoABuilder) WithParams(params map[string]any) *PoABuilder {
	for k, v := range params {
		b.params[k] = v
	}
	return b
}

// WithParam adds a single action parameter.
func (b *PoABuilder) WithParam(key string, value any) *PoABuilder {
	b.params[key] = value
	return b
}

// WithConstraint adds a constraint.
func (b *PoABuilder) WithConstraint(key string, value any) *PoABuilder {
	b.constraints[key] = value
	return b
}

// WithConstraints adds multiple constraints.
func (b *PoABuilder) WithConstraints(constraints map[string]any) *PoABuilder {
	for k, v := range constraints {
		b.constraints[k] = v
	}
	return b
}

// Legal sets the legal grounding.
func (b *PoABuilder) Legal(leg LegalGrounding) *PoABuilder {
	b.leg = &leg
	return b
}

// TTL sets the time-to-live (default: 5 minutes).
func (b *PoABuilder) TTL(d time.Duration) *PoABuilder {
	b.ttl = d
	return b
}

// Issuer sets the issuer SPIFFE ID.
func (b *PoABuilder) Issuer(spiffeID string) *PoABuilder {
	b.iss = spiffeID
	return b
}

// Audience sets the audience SPIFFE ID(s).
func (b *PoABuilder) Audience(spiffeIDs ...string) *PoABuilder {
	b.aud = spiffeIDs
	return b
}

// Build creates the PoA mandate.
func (b *PoABuilder) Build() (*PoA, error) {
	if b.sub == "" {
		return nil, errors.New("subject (agent SPIFFE ID) is required")
	}
	if b.act == "" {
		return nil, errors.New("action is required")
	}
	if b.leg == nil {
		return nil, errors.New("legal grounding is required")
	}

	now := time.Now().Unix()
	poa := &PoA{
		Sub: b.sub,
		Act: b.act,
		Con: Constraints{
			Params:      b.params,
			Constraints: b.constraints,
		},
		Leg: *b.leg,
		Iat: now,
		Exp: now + int64(b.ttl.Seconds()),
		Jti: "poa_" + uuid.New().String()[:12],
		Iss: b.iss,
		Aud: b.aud,
	}

	if err := poa.Validate(); err != nil {
		return nil, err
	}

	return poa, nil
}

// MustBuild is like Build but panics on error.
func (b *PoABuilder) MustBuild() *PoA {
	poa, err := b.Build()
	if err != nil {
		panic(err)
	}
	return poa
}
